src/nms-api/IMessageProducer.cs

```diff
@@ -54,6 +58,48 @@ public interface IMessageProducer : System.IDisposable
 		/// </summary>
 		void Send(IDestination destination, IMessage message, MsgDeliveryMode deliveryMode, MsgPriority priority, TimeSpan timeToLive);
 
+                /// <summary>
+                /// Sends the message to the default destination for this producer
+                /// </summary>
+                void Send(IMessage message, CompletionListener completionListener);
+
+                /// <summary>
+                /// Sends the message to the default destination with the explicit QoS configuration
+                /// </summary>
+                void Send(IMessage message, MsgDeliveryMode deliveryMode, MsgPriority priority, TimeSpan timeToLive, CompletionListener completionListener);
+
+                /// <summary>
+                /// Sends the message to the given destination
+                /// </summary>
+                void Send(IDestination destination, IMessage message, CompletionListener completionListener);
+
+                /// <summary>
+                /// Sends the message to the given destination with the explicit QoS configuration
+                /// </summary>
+                void Send(IDestination destination, IMessage message, MsgDeliveryMode deliveryMode, MsgPriority priority, TimeSpan timeToLive, CompletionListener completionListener);
+
+
+                /// <summary>
+                /// Sends the message to the default destination for this producer
+                /// </summary>
+                Task SendAsync(IMessage message);
+
+                /// <summary>
+                /// Sends the message to the default destination with the explicit QoS configuration
+                /// </summary>
+                Task SendAsync(IMessage message, MsgDeliveryMode deliveryMode, MsgPriority priority, TimeSpan timeToLive);
+
+                /// <summary>
+                /// Sends the message to the given destination
+                /// </summary>
+                Task SendAsync(IDestination destination, IMessage message);
+
+                /// <summary>
+                /// Sends the message to the given destination with the explicit QoS configuration
+                /// </summary>
+                Task SendAsync(IDestination destination, IMessage message, MsgDeliveryMode deliveryMode, MsgPriority priority, TimeSpan timeToLive);
```

Comment:
In order to support Asynchronous API in that way we need to drop support for net35, as TPL (Task Parallel Library) was introduced in net40.

Reply:
yes, we will have to, i think that's fine, before we merge this, we should make a 1.x branch for any maintenance needs on the older 1.x